src/cpu/flexcpu/simple_dataflow_thread.cc

```diff
@@ -969,43 +980,136 @@ SDCPUThread::populateDependencies(shared_ptr<InflightInst> inst_ptr)
 
     // BEGIN Sequential Consistency
 
-    // TODO IMPORTANT this still doesn't make it safe to speculate on memory
-    //      instructions, since if we produce a fault and need to squash, we
-    //      still need a way to inform the cpu to undo stores being sent to
-    //      memory.
-
     // By creating a dependency between any memory requests and the most recent
     // prior memory request, all memory instructions are enforced to execute in
     // program-order.
-    if (static_inst->isMemRef() || static_inst->isMemBarrier()) {
-        for (auto itr = inflightInsts.rbegin();
-             itr != inflightInsts.rend();
-             ++itr) {
-            if (*itr == inst_ptr) continue;
-            if ((*itr)->isComplete() || ((*itr)->isSquashed())) continue;
+    // if (static_inst->isMemRef() || static_inst->isMemBarrier()) {
+    //     for (auto itr = inflightInsts.rbegin();
+    //          itr != inflightInsts.rend();
+    //          ++itr) {
+    //         if (*itr == inst_ptr) continue;
+    //         if ((*itr)->isComplete() || ((*itr)->isSquashed())) continue;
+
+    //         const StaticInstPtr other_si = (*itr)->staticInst();
+
+    //         if (other_si->isMemRef() || other_si->isMemBarrier()) {
+    //             DPRINTF(SDCPUDeps, "Dep %d -> %d [mem barrier]\n",
+    //                     inst_ptr->seqNum(), (*itr)->seqNum());
+    //             inst_ptr->addMemDependency(*itr);
+    //             break;
+    //         }
+    //     }
+    // }
 
-            const StaticInstPtr other_si = (*itr)->staticInst();
+    // END Sequential Consistency
 
-            if (other_si->isMemRef() || other_si->isMemBarrier()) {
-                DPRINTF(SDCPUDeps, "Dep %d -> %d [mem barrier]\n",
-                        inst_ptr->seqNum(), (*itr)->seqNum());
-                inst_ptr->addDependency(*itr);
-                break;
+    // BEGIN Conservative Memory dependence ordering
+
+    if (static_inst->isLoad()) {
+        // TODO check all previous memory referencing instructions.
+        //      if effaddr resolved, add self check to see if memory needs
+        //         dependence.
+        //      else add effaddr dependence.
+        weak_ptr<InflightInst> weak_inst = inst_ptr;
+
+        for (shared_ptr<InflightInst>& other : inflightInsts) {
```

Comment:
Remind me what order this is (in a comment). Are we starting with the most youngest or oldest instruction?

Reply:
front is oldest, so we are going from oldest to youngest in this case. Order doesn't matter in this case though, because we are checking all older stores.